ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,575 @@
+      that method invocation is actually, say, activity invocation.
+
+## Dynamic Workflows, Signals, Queries, and Updates
+
+* Support dynamic pieces similar to how we have done in Python and .NET.
+* Create a new `Temporalio::Converters::RawValue` object that is a pass-through object for raw values (still subject to
+  payload codecs), and a `payload_converter` is available in workflow/activity context to convert it.
+  * Handler methods will use `RawValue` for args if told to, similar with workflow `execute`/`initialize` if
+    `workflow_raw_args` is set.
+  * `RawValue` can be returned too.
+* To mark a workflow as dynamic, `workflow_dynamic` class method is called. Only one dynamic workflow can be present for
+  a worker. The workflow can retrieve its workflow type name via `Workflow.info`
+* Dynamic signals, queries, and updates are enabled via `dynamic: true` on the calls before the methods.
+  * Dynamic handlers first argument is the string name (it's not a symbol), and all successive arguments are the actual
+    arguments.
+  * Only one dynamic handler of a type (signal, query, or update) can be on a workflow.
+
+## Workflow Asynchrony
+
+Ruby does not have a built-in task/promise/future/etc concept. However it does have fibers and we can leverage those.
+
+### Custom Fiber Scheduler
+
+We will write a custom Fiber scheduler. This will provide the following benefits:
+
+* Can leverage `Timeout` for timeouts.
+* Can leverage `sleep` for sleeps.
+* Able to fail some IO calls.
+* Allows usage of most of https://github.com/socketry/async.
+  * We will have many tests confirming many behaviors here but we still get to avoid a runtime dependency.
+  * ❓ Why not just take a dependency on `async` gem? It does have a deterministic scheduler and does a decent job of
+    everything we want, but not only do we want to avoid transitive dependencies in our library (for versioning and
+    other reasons), we cannot rely on its logic into the future.
+
+### Futures
+
+Since Ruby does not have a concept of a task/promise/future, we have to create one. Users need all the common features
+of a future without us depending on a library. Here's a high-level view of what the class may look like:
+
+```ruby
+class Future
+  # Returns [Future, Setter].
+  def new_with_setter
+
+  # Returns future whose result is set to the first completed future's result or
+  # failed with the first future's failure.
+  def self.any_of(*futures)
+  # Returns future whose result is set to nil when all futures succeed or
+  # has its failure set to the first failure.
+  def self.all_of(*futures)
+
+  # Returns future whose result is set to the first future completed (so a
+  # future in a future).
+  def self.any_of_no_raise(*futures)
+  # Returns future whose result is set to nil when all futures complete
+  # regardless of whether any future failed.
+  def self.all_of_no_raise(*futures)
```

Comment:
> because what's the difference between `any_of_no_raise(...).wait` and `any_of_no_raise(...).wait_no_raise`?

No big difference if you have both of these calls together. But they have value independently (the first one returns a future the second you have chosen not to even be able to look at the failure) and have value as consistent shortcuts. I can't imagine anyone would call the latter combo, `wait_no_raise` will more often be used for people using singular futures they don't care about.

Most languages have forms of these. But I am open to better names though don't want to get too inconsistent with each other and don't want these to be arguments since they change the return type. Basic requirements I think are:

* Ability to wait for first of multiple futures or raise on first failure
* Ability to wait for first of multiple futures but do not raise on first failure
* Ability to wait on all of multiple futures or raise on first failure
* Ability to wait on all of multiple futures but do not raise on first failure
* Ability to wait on a future or raise on failure
* Ability to wait on a future and not raise on failure

This was my best attempt after many permutations to get those 6 methods in. Definitely open to clearer suggestions.

Reply:
Yeah I would then change the names a-la my suggestions - the most important part I think is just not to use `no_raise` in these names since it's not _really_ about not raising, it's about not short-circuiting in the event one fails.

Reply:
> Yeah I would then change the names a-la my suggestions 

Do you have a specific suggestion on names? So instead of:

* `any_of`
* `any_of_no_raise`
* `all_of`
* `all_of_no_raise`
* `wait`
* `wait_no_raise`

reading comment above, something like:

* `any_of`
* `any_of_success` - says success but includes failure?
* `try_all_of` - fails future if any fail?
* `all_of` - does not fail future if any fail, so one "x_of" doesn't ail the future and one does?
* `wait`
* `wait_no_raise`

Maybe something like `try` instead of `_no_raise`, e.g.

* `any_of`
* `try_any_of`
* `all_of`
* `try_all_of`
* `wait`
* `try_wait`

Reply:
I like the try names, but I think they should actually be the short-circuiting versions. IE: Existing `any_of` and `all_of` become the `try_xxx` versions, and the current `_no_raise` versions just become `any_of` and `all_of`. Since "try" implies "try until something breaks". This also matches the Rust naming scheme for these, FWIW.

I would probably keep `wait_no_raise` though rather than `try` there just because that's more obvious. And, I think, to my point, that is actually about raising or not where the other stuff is about short-circuiting.

Reply:
Hrmm, to me "try" means failure is captured, e.g. .NET `int.Parse` vs `int.TryParse`. One fails as normal, another captures failure (same as the actual `try` statement in many languages).

Reply:
The `try` here is like "try them, but fail if any fails soon" whereas without it is "definitely do all the futures even if some of them fail".

But, I wouldn't cry about having it the other way around, I think that can work too w/ clear docs. I do think `wait` and `wait_no_raise` should stay that way though.

Reply:
Sure, so `wait`, `wait_no_raise` settled. Any other term we can think of for "swallow error" form of the others besides `try`? I will update proposal with the `try_` form, but definitely open to other words here.

Reply:
Proposal updated to change `any_of_no_raise` and `all_of_no_raise` to `try_any_of` and `try_all_of` and clarify they do not raise when `wait`ed on.